pkg/ibmvpc/subnetsIPBlocks.go

```diff
@@ -0,0 +1,252 @@
+//       when a free address is needed, we take the first address of the block and remove it from the free blocks list.
+
+// /////////////////////////////////////////////////////////////////////////////////////////
+// subnetIPBlocks hold the blocks of a subnet
+type subnetIPBlocks struct {
+	subnetOriginalBlock  *ipblock.IPBlock // the block of the original cidr of the subnet
+	splitByFiltersBlocks []*ipblock.IPBlock
+	freeAddressesBlocks  []*ipblock.IPBlock // each block in splitByFiltersBlocks has a corresponding block at freeAddressesBlocks
+	fullyReservedBlocks  []bool             // true if all the addresses in the block are reserved IP
+}
+type subnetsIPBlocks map[string]*subnetIPBlocks
+
+func getSubnetsIPBlocks(rc *datamodel.ResourcesContainerModel) (subnetsBlocks subnetsIPBlocks, err error) {
+	subnetsBlocks = subnetsIPBlocks{}
+	// get all the original blocks of the subnets:
+	if err := subnetsBlocks.getSubnetsOriginalBlocks(rc); err != nil {
+		return nil, err
+	}
+	// calc the filters blocks:
+	filtersBlocks, err := getFiltersBlocks(rc)
+	if err != nil {
+		return nil, err
+	}
+	// calc the splitByFiltersBlocks:
+	subnetsBlocks.splitSubnetsOriginalBlocks(rc, filtersBlocks)
+	// calc the freeAddressesBlocks:
+	if err := subnetsBlocks.getSubnetsFreeBlocks(rc); err != nil {
+		return nil, err
+	}
+	return subnetsBlocks, nil
+}
+
+func (subnetsBlocks subnetsIPBlocks) getSubnetsOriginalBlocks(rc *datamodel.ResourcesContainerModel) (err error) {
+	for _, subnetObj := range rc.SubnetList {
+		subnetsBlocks[*subnetObj.CRN] = &subnetIPBlocks{}
+		subnetsBlocks[*subnetObj.CRN].subnetOriginalBlock, err = ipblock.FromCidr(*subnetObj.Ipv4CIDRBlock)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// splitSubnetsOriginalBlocks():
+// the goal of this func is to split the original subnet cidr to disjoint blocks, according to the filters blocks:
+// for each subnet it gets all the blocks that intersect with subnet original block
+func (subnetsBlocks subnetsIPBlocks) splitSubnetsOriginalBlocks(rc *datamodel.ResourcesContainerModel, filtersBlocks filtersBlocks) {
+	for _, subnetObj := range rc.SubnetList {
+		subnetsBlocks[*subnetObj.CRN].splitByFiltersBlocks =
+			splitSubnetOriginalBlock(subnetsBlocks[*subnetObj.CRN].subnetOriginalBlock, filtersBlocks[*subnetObj.VPC.CRN])
+	}
+}
+
+func splitSubnetOriginalBlock(subnetOriginalBlock *ipblock.IPBlock, filtersBlocks []*ipblock.IPBlock) []*ipblock.IPBlock {
+	filtersBlocksOnSubnet := []*ipblock.IPBlock{}
+	for _, filterBlock := range filtersBlocks {
+		filterBlocksOnSubnet := subnetOriginalBlock.Intersect(filterBlock)
+		if !filterBlocksOnSubnet.IsEmpty() {
+			filtersBlocksOnSubnet = append(filtersBlocksOnSubnet, filterBlocksOnSubnet)
+		}
+	}
+	return filtersBlocksOnSubnet
+}
+
+// getSubnetsFreeBlocks() - calc all the addresses that are not allocated:
+// for each subnet:
+//  1. make a copy of the splitByFiltersBlocks
+//  2. remove the addresses that was already allocated
```

Comment:
```suggestion
//  2. remove the addresses that were already allocated from the copy [refer to the copy, perhaps use another name]
```

Reply:
Please make the documentation easier to understand; you make a copy of the disjoint filter atomic blocks, and then remove from each such block all the already allocated addresses, thus resulting in the free addresses in each

Reply:
fixed